CalculateEligibilityDeterminationJob is failing in several environments

**Describe the bug**
Error message:

```
A durable job has failed repeatedly on civiformtest.seattle.gov
This needs to be investigated by IT staff or the CiviForm core team (civiform-technical@googlegroups.com).
Error report for: job_name="CALCULATE_ELIGIBILITY_DETERMINATION_JOB", job_ID=635
Eligibility Determination: stopping early after 10 errors
END_ERROR
JobRunner_JobFailed ExecutionException job_name="CALCULATE_ELIGIBILITY_DETERMINATION_JOB", job_ID=635, attempts_remaining=2, duration_s=4.628000, error_message=jakarta.persistence.PersistenceException: java.sql.SQLException: Connection is closed, trace=java.util.concurrent.ExecutionException: jakarta.persistence.PersistenceException: java.sql.SQLException: Connection is closed
at java.base/java.util.concurrent.CompletableFuture.reportGet(Unknown Source)
at java.base/java.util.concurrent.CompletableFuture.get(Unknown Source)
at durablejobs.AbstractDurableJobRunner.runJobWithTimeout(AbstractDurableJobRunner.java:230)
at durablejobs.AbstractDurableJobRunner.runJob(AbstractDurableJobRunner.java:142)
at durablejobs.AbstractDurableJobRunner.runJobs(AbstractDurableJobRunner.java:90)
at modules.DurableJobModule$DurableJobRunnerScheduler.lambda$new$0(DurableJobModule.java:102)
at org.apache.pekko.actor.LightArrayRevolverScheduler$$anon$1$$anon$2.run(LightArrayRevolverScheduler.scala:124)
at org.apache.pekko.dispatch.TaskInvocation.run(AbstractDispatcher.scala:59)
at org.apache.pekko.dispatch.ForkJoinExecutorConfigurator$PekkoForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:57)
at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: jakarta.persistence.PersistenceException: java.sql.SQLException: Connection is closed
at io.ebeaninternal.server.transaction.JdbcTransaction.doRollback(JdbcTransaction.java:1024)
at io.ebeaninternal.server.transaction.JdbcTransaction.rollback(JdbcTransaction.java:1009)
at io.ebeaninternal.server.transaction.JdbcTransaction.rollback(JdbcTransaction.java:996)
at durablejobs.jobs.CalculateEligibilityDeterminationJob.run(CalculateEligibilityDeterminationJob.java:111)
at durablejobs.AbstractDurableJobRunner.lambda$runJobWithTimeout$0(AbstractDurableJobRunner.java:221)
at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(Unknown Source)
at play.core.j.ClassLoaderExecutionContext.$anonfun$execute$1(ClassLoaderExecutionContext.scala:64)
at org.apache.pekko.dispatch.TaskInvocation.run(AbstractDispatcher.scala:59)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.sql.SQLException: Connection is closed
at com.zaxxer.hikari.pool.ProxyConnection$ClosedConnection.lambda$getClosedConnection$0(ProxyConnection.java:502)
at jdk.proxy3/jdk.proxy3.$Proxy42.rollback(Unknown Source)
at com.zaxxer.hikari.pool.ProxyConnection.rollback(ProxyConnection.java:385)
at com.zaxxer.hikari.pool.HikariProxyConnection.rollback(HikariProxyConnection.java)
at io.ebeaninternal.server.transaction.JdbcTransaction.performRollback(JdbcTransaction.java:812)
at io.ebeaninternal.server.transaction.JdbcTransaction.doRollback(JdbcTransaction.java:1022)
... 10 more
END_ERROR
Eligibility Determination: stopping early after 10 errors
END_ERROR
JobRunner_JobFailed ExecutionException job_name="CALCULATE_ELIGIBILITY_DETERMINATION_JOB", job_ID=635, attempts_remaining=1, duration_s=1.509000, error_message=jakarta.persistence.PersistenceException: java.sql.SQLException: Connection is closed, trace=java.util.concurrent.ExecutionException: jakarta.persistence.PersistenceException: java.sql.SQLException: Connection is closed
at java.base/java.util.concurrent.CompletableFuture.reportGet(Unknown Source)
at java.base/java.util.concurrent.CompletableFuture.get(Unknown Source)
at durablejobs.AbstractDurableJobRunner.runJobWithTimeout(AbstractDurableJobRunner.java:230)
at durablejobs.AbstractDurableJobRunner.runJob(AbstractDurableJobRunner.java:142)
at durablejobs.AbstractDurableJobRunner.runJobs(AbstractDurableJobRunner.java:90)
at modules.DurableJobModule$DurableJobRunnerScheduler.lambda$new$0(DurableJobModule.java:102)
at org.apache.pekko.actor.LightArrayRevolverScheduler$$anon$1$$anon$2.run(LightArrayRevolverScheduler.scala:124)
at org.apache.pekko.dispatch.TaskInvocation.run(AbstractDispatcher.scala:59)
at org.apache.pekko.dispatch.ForkJoinExecutorConfigurator$PekkoForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:57)
at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: jakarta.persistence.PersistenceException: java.sql.SQLException: Connection is closed
at io.ebeaninternal.server.transaction.JdbcTransaction.doRollback(JdbcTransaction.java:1024)
at io.ebeaninternal.server.transaction.JdbcTransaction.rollback(JdbcTransaction.java:1009)
at io.ebeaninternal.server.transaction.JdbcTransaction.rollback(JdbcTransaction.java:996)
at durablejobs.jobs.CalculateEligibilityDeterminationJob.run(CalculateEligibilityDeterminationJob.java:111)
at durablejobs.AbstractDurableJobRunner.lambda$runJobWithTimeout$0(AbstractDurableJobRunner.java:221)
at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(Unknown Source)
at play.core.j.ClassLoaderExecutionContext.$anonfun$execute$1(ClassLoaderExecutionContext.scala:64)
at org.apache.pekko.dispatch.TaskInvocation.run(AbstractDispatcher.scala:59)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.sql.SQLException: Connection is closed
at com.zaxxer.hikari.pool.ProxyConnection$ClosedConnection.lambda$getClosedConnection$0(ProxyConnection.java:502)
at jdk.proxy3/jdk.proxy3.$Proxy42.rollback(Unknown Source)
at com.zaxxer.hikari.pool.ProxyConnection.rollback(ProxyConnection.java:385)
at com.zaxxer.hikari.pool.HikariProxyConnection.rollback(HikariProxyConnection.java)
at io.ebeaninternal.server.transaction.JdbcTransaction.performRollback(JdbcTransaction.java:812)
at io.ebeaninternal.server.transaction.JdbcTransaction.doRollback(JdbcTransaction.java:1022)
... 10 more
END_ERROR
Eligibility Determination: stopping early after 10 errors
END_ERROR
JobRunner_JobFailed ExecutionException job_name="CALCULATE_ELIGIBILITY_DETERMINATION_JOB", job_ID=635, attempts_remaining=0, duration_s=1.182000, error_message=jakarta.persistence.PersistenceException: java.sql.SQLException: Connection is closed, trace=java.util.concurrent.ExecutionException: jakarta.persistence.PersistenceException: java.sql.SQLException: Connection is closed
at java.base/java.util.concurrent.CompletableFuture.reportGet(Unknown Source)
at java.base/java.util.concurrent.CompletableFuture.get(Unknown Source)
at durablejobs.AbstractDurableJobRunner.runJobWithTimeout(AbstractDurableJobRunner.java:230)
at durablejobs.AbstractDurableJobRunner.runJob(AbstractDurableJobRunner.java:142)
at durablejobs.AbstractDurableJobRunner.runJobs(AbstractDurableJobRunner.java:90)
at modules.DurableJobModule$DurableJobRunnerScheduler.lambda$new$0(DurableJobModule.java:102)
at org.apache.pekko.actor.LightArrayRevolverScheduler$$anon$1$$anon$2.run(LightArrayRevolverScheduler.scala:124)
at org.apache.pekko.dispatch.TaskInvocation.run(AbstractDispatcher.scala:59)
at org.apache.pekko.dispatch.ForkJoinExecutorConfigurator$PekkoForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:57)
at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: jakarta.persistence.PersistenceException: java.sql.SQLException: Connection is closed
at io.ebeaninternal.server.transaction.JdbcTransaction.doRollback(JdbcTransaction.java:1024)
at io.ebeaninternal.server.transaction.JdbcTransaction.rollback(JdbcTransaction.java:1009)
at io.ebeaninternal.server.transaction.JdbcTransaction.rollback(JdbcTransaction.java:996)
at durablejobs.jobs.CalculateEligibilityDeterminationJob.run(CalculateEligibilityDeterminationJob.java:111)
at durablejobs.AbstractDurableJobRunner.lambda$runJobWithTimeout$0(AbstractDurableJobRunner.java:221)
at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(Unknown Source)
at play.core.j.ClassLoaderExecutionContext.$anonfun$execute$1(ClassLoaderExecutionContext.scala:64)
at org.apache.pekko.dispatch.TaskInvocation.run(AbstractDispatcher.scala:59)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
at java.base/java.lang.Thread.run(Unknown Source) 
```

**To Reproduce**
Steps to reproduce the behavior:

1. Create a program with an eligibility condition and publish
2. Apply to the program as an applicant
3. Update the data in the db so that the `eligibility_determination` column on the `applications` table equals `NOT_COMPUTED`
4. Run the `CalculateEligibilityDeterminationJob` via the dev tools and see the error
5. See that `eligibility_determination` column on the `applications` table is still `NOT_COMPUTED`

**Expected behavior**
`CALCULATE_ELIGIBILITY_DETERMINATION_JOB` should run without error and all `NOT_COMPUTED` values in the `eligibility_determination` column should be updated to the [appropriate value](https://github.com/civiform/civiform/blob/main/server/app/models/EligibilityDetermination.java).


- Device: [e.g. iPhone6]
- OS: [e.g. iOS8.1]
- Browser [e.g. stock browser, safari]
- Version [e.g. 22]

**Additional context**
Add any other context about the problem here.

--- a/server/app/durablejobs/jobs/CalculateEligibilityDeterminationJob.java
+++ b/server/app/durablejobs/jobs/CalculateEligibilityDeterminationJob.java
@@ -12,12 +12,13 @@ import javax.inject.Inject;
 import models.ApplicationModel;
 import models.EligibilityDetermination;
 import models.PersistedDurableJobModel;
-import models.ProgramModel;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import services.applicant.ApplicantService;
 import services.applicant.ReadOnlyApplicantProgramService;
 import services.program.ProgramDefinition;
+import services.program.ProgramNotFoundException;
+import services.program.ProgramService;
 
 /**
  * Calculate the eligibility determination for applications submitted before the pre-compute
@@ -27,14 +28,18 @@ public final class CalculateEligibilityDeterminationJob extends DurableJob {
   private static final Logger logger =
       LoggerFactory.getLogger(CalculateEligibilityDeterminationJob.class);
   private final ApplicantService applicantService;
+  private final ProgramService programService;
 
   private final Database database;
   private final PersistedDurableJobModel persistedDurableJobModel;
 
   @Inject
   public CalculateEligibilityDeterminationJob(
-      ApplicantService applicantService, PersistedDurableJobModel persistedDurableJobModel) {
+      ApplicantService applicantService,
+      ProgramService programService,
+      PersistedDurableJobModel persistedDurableJobModel) {
     this.applicantService = checkNotNull(applicantService);
+    this.programService = checkNotNull(programService);
     this.persistedDurableJobModel = checkNotNull(persistedDurableJobModel);
     this.database = DB.getDefault();
   }
@@ -47,7 +52,6 @@ public final class CalculateEligibilityDeterminationJob extends DurableJob {
   @Override
   public void run() {
     logger.info("Starting job to calculate eligibility determination.");
-
     try (Transaction jobTransaction = database.beginTransaction(TxIsolation.SERIALIZABLE)) {
       jobTransaction.setBatchMode(true);
       jobTransaction.setBatchSize(50);
@@ -72,11 +76,14 @@ public final class CalculateEligibilityDeterminationJob extends DurableJob {
                   "Calculating eligibility determination for application id {}",
                   application.id); // Use the captured ID
 
-              ProgramModel pm = application.getProgram();
-              ProgramDefinition programDefinition = pm.getProgramDefinition();
+              Long programId = application.getProgram().id;
+              ProgramDefinition programDefinition =
+                  programService.getFullProgramDefinition(programId);
+
               ReadOnlyApplicantProgramService roAppProgramService =
                   applicantService.getReadOnlyApplicantProgramService(
                       application, programDefinition);
+
               EligibilityDetermination eligibilityDetermination =
                   applicantService.calculateEligibilityDetermination(
                       programDefinition, roAppProgramService);
@@ -86,7 +93,7 @@ public final class CalculateEligibilityDeterminationJob extends DurableJob {
               logger.warn("Query returned a null application despite hasNext() being true.");
               errorCount++;
             }
-          } catch (RuntimeException e) {
+          } catch (RuntimeException | ProgramNotFoundException e) {
             errorCount++;
             final Long finalApplicationId = currentApplicationId;
             applicationOptional.ifPresentOrElse(

--- a/server/app/modules/DurableJobModule.java
+++ b/server/app/modules/DurableJobModule.java
@@ -40,6 +40,7 @@ import repository.VersionRepository;
 import scala.concurrent.ExecutionContext;
 import services.applicant.ApplicantService;
 import services.cloud.PublicStorageClient;
+import services.program.ProgramService;
 
 /**
  * Configures {@link durablejobs.DurableJob}s with their {@link DurableJobName} and, if they are
@@ -115,6 +116,7 @@ public final class DurableJobModule extends AbstractModule {
   public DurableJobRegistry provideRecurringDurableJobRegistry(
       AccountRepository accountRepository,
       ApplicantService applicantService,
+      ProgramService programService,
       @BindingAnnotations.Now Provider<LocalDateTime> nowProvider,
       PersistedDurableJobRepository persistedDurableJobRepository,
       PublicStorageClient publicStorageClient,
@@ -155,7 +157,8 @@ public final class DurableJobModule extends AbstractModule {
         DurableJobName.CALCULATE_ELIGIBILITY_DETERMINATION_JOB,
         JobType.RECURRING,
         persistedDurableJob ->
-            new CalculateEligibilityDeterminationJob(applicantService, persistedDurableJob),
+            new CalculateEligibilityDeterminationJob(
+                applicantService, programService, persistedDurableJob),
         new RecurringJobExecutionTimeResolvers.Sunday2Am());
 
     return durableJobRegistry;

--- a/server/test/durablejobs/jobs/CalculateEligibilityDeterminationTest.java
+++ b/server/test/durablejobs/jobs/CalculateEligibilityDeterminationTest.java
@@ -34,6 +34,7 @@ import services.applicant.ReadOnlyApplicantProgramService;
 import services.applicant.question.Scalar;
 import services.applications.ApplicationService;
 import services.program.ProgramDefinition;
+import services.program.ProgramService;
 import services.question.QuestionService;
 import services.question.types.NameQuestionDefinition;
 import services.question.types.QuestionDefinitionConfig;
@@ -46,6 +47,7 @@ public class CalculateEligibilityDeterminationTest extends ResetPostgres {
   private VersionRepository versionRepository;
   private QuestionService questionService;
   ApplicationService applicationService = instanceOf(ApplicationService.class);
+  ProgramService programService = instanceOf(ProgramService.class);
   ApplicationModel applicationModel = instanceOf(ApplicationModel.class);
   CiviFormProfile trustedIntermediaryProfile;
   ProgramDefinition programDefinition;
@@ -93,7 +95,7 @@ public class CalculateEligibilityDeterminationTest extends ResetPostgres {
     application.setEligibilityDetermination(EligibilityDetermination.NO_ELIGIBILITY_CRITERIA);
 
     CalculateEligibilityDeterminationJob job =
-        new CalculateEligibilityDeterminationJob(applicantService, jobModel);
+        new CalculateEligibilityDeterminationJob(applicantService, programService, jobModel);
     job.run();
     application.refresh();
 
@@ -111,7 +113,7 @@ public class CalculateEligibilityDeterminationTest extends ResetPostgres {
         .isEqualTo(EligibilityDetermination.NOT_COMPUTED);
 
     CalculateEligibilityDeterminationJob job =
-        new CalculateEligibilityDeterminationJob(applicantService, jobModel);
+        new CalculateEligibilityDeterminationJob(applicantService, programService, jobModel);
     job.run();
     application.refresh();
 
@@ -139,7 +141,7 @@ public class CalculateEligibilityDeterminationTest extends ResetPostgres {
         .isEqualTo(EligibilityDetermination.NOT_COMPUTED);
 
     CalculateEligibilityDeterminationJob job =
-        new CalculateEligibilityDeterminationJob(applicantService, jobModel);
+        new CalculateEligibilityDeterminationJob(applicantService, programService, jobModel);
     job.run();
     firstApp.refresh();
     secondApp.refresh();
@@ -202,7 +204,7 @@ public class CalculateEligibilityDeterminationTest extends ResetPostgres {
         .isEqualTo(EligibilityDetermination.NOT_COMPUTED);
 
     CalculateEligibilityDeterminationJob job =
-        new CalculateEligibilityDeterminationJob(applicantService, jobModel);
+        new CalculateEligibilityDeterminationJob(applicantService, programService, jobModel);
     job.run();
     application.refresh();
 
@@ -233,7 +235,7 @@ public class CalculateEligibilityDeterminationTest extends ResetPostgres {
             any(ProgramDefinition.class), any(ReadOnlyApplicantProgramService.class));
 
     CalculateEligibilityDeterminationJob job =
-        new CalculateEligibilityDeterminationJob(applicantService, jobModel);
+        new CalculateEligibilityDeterminationJob(applicantService, programService, jobModel);
     job.run();
 
     applicationsToProcess.forEach(ApplicationModel::refresh);

